### PR TITLE
UmdTask74_Fall2025_xgboost_Employee_Attrition_Prediction

### PR DESCRIPTION
**Pull request by tharun2k1**
Issue: #74 

NAME: Tharunkumar Chandrasekaran
UID:121323101

**Objective**  
Predict whether employees are likely to leave a company using demographic, job satisfaction, and performance data.

**Dataset**  
IBM HR Analytics Employee Attrition & Performance (Kaggle), automatically downloaded and managed using `kagglehub`.

**What this project does**

- Implements an end-to-end machine learning pipeline to predict employee attrition.
- Automates dataset download and loading via **KaggleHub** (no manual CSV handling needed).
- Performs data preprocessing using **scikit-learn Pipelines** and **ColumnTransformer**:
  - One-hot encodes categorical features.
  - Scales numeric features.
- Trains and evaluates multiple models:
  - **XGBoost** (primary model, with class weighting for imbalance).
  - **Logistic Regression** (baseline, `class_weight="balanced"`).
  - **Random Forest** (baseline, `class_weight="balanced"`).
- Compares models using:
  - **Accuracy**
  - **F1-score** (for the attrition = Yes class)
  - **ROC-AUC**
- Explores **threshold tuning** for XGBoost to trade off F1 vs recall for the minority “Attrition = Yes” class.
- Generates **feature importance** and **model explanation** plots (XGBoost feature importance, SHAP-style interpretation in the notebook).
- Organizes everything in a main Jupyter notebook for:
  - Exploratory Data Analysis (EDA)
  - Model training and evaluation
  - Result interpretation and summary.

## Folder Structure 

Fall2025_xgboost_Employee_Attrition_Prediction/
|
├── docker_build.sh                # Build Docker image
├── docker_bash.sh                 # Optional: open bash shell inside container
├── docker_jupyter.sh             # Run Jupyter Notebook in container
├── Dockerfile                          # Image definition (Python 3.10 + libs + Jupyter)
├── README.md                      # How to build/run in Docker + project overview
├── requirements.txt               # Python dependencies
├── xgboost.API.ipynb            # Small notebook showing XGBoost API + wrapper usage
├── xgboost.API.md                # Text documentation of XGBoost API + wrapper
├── xgboost.example.ipynb    # MAIN notebook: full E2E project (EDA + models)
└── xgboost.example.md       # Text summary of the example and results

 ## How to Run (Docker + Jupyter)

Full step-by-step instructions are in **`README.md`** inside this folder.  
Quick summary for the grader:

1. **Clone the repository**

```bash
git clone https://github.com/gpsaggese-org/umd_classes.git
cd umd_classes
```

2. **Go to the project folder**

```bash
cd class_project/MSML610/Fall2025/Projects/Fall2025_xgboost_Employee_Attrition_Prediction
```

3. **Make the helper scripts executable (first time only)**

```bash
chmod +x docker_build.sh docker_bash.sh docker_jupyter.sh
```

4. **Build the Docker image**

```bash
./docker_build.sh
```

5. **Launch Jupyter Notebook inside Docker**

```bash
./docker_jupyter.sh
```

6. **Open the notebook in a browser**

- Go to: `http://127.0.0.1:8888/tree`
- Open **`xgboost.example.ipynb`**
- Run the notebook top-to-bottom to reproduce:
  - EDA
  - XGBoost training & evaluation
  - Model comparison
  - Threshold tuning
  - Interpretability and clustering analysis

---

### CODE REVIEW

**Comment by aver81** [comment]
@tharun2k1 code looks good. just fix the file name, it should be as per convention

**Comment by aver81** [comment]
@tharun2k1  remember to do a git pull before pushing changes. I pushed some changes to fix conflicts

**Comment by aver81** [comment]
@tharun2k1 

1. README.md looks good
2. Dockerfile looks good
3. API and example files look good

Everything looks good. Feel free to make any other changes you want, but overall it looks good
